### ISSUE TITLE
Document and provide alternatives to `ddev share`

### ISSUE THREAD

**Issue by rfay** [ddev/ddev#6441]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

`ddev share` is great, but it's tied to ngrok, and it has very specific use-cases

### Describe your solution

Options include:
* Replace `share` with a custom command so that people could understand better how to override it
* Handle signals better so that shutdown shuts down the connection better
* Add alternate share techniques like `cloudflared` that are more general, help people understand how to use them.

### Additional context

* https://github.com/ddev/ddev/issues/4962
* https://github.com/ddev/ddev/issues/4365

**Comment by tyler36** [comment]
Ideally, I'd like to see "share" move to a global command.
This would:

- allow developers to take control and overwrite as required
- allow per-project overrides
- maintain backwards compatibility and the standard "ddev share" command across all DDEV projects
- provide centralized place for "aliases" (#6434) if people wanted to `ddev ngrok`, `ddev <newer-share-service>`

**Comment by rfay** [comment]
Good argument. A global shell command would allow that. It also could have args like `--ngrok` or `--cloudflare` etc

**Comment by stasadev** [comment]
List of possible alternatives:

- https://github.com/anderspitman/awesome-tunneling

**Comment by rfay** [comment]
Example of using cloudflared and ngrok in a more complex environment: https://stackoverflow.com/questions/78729503/drupal-10-multisite-using-ddev-share-or-cloudflare/78744731#78744731

**Comment by joelhans** [comment]
Love seeing ngrok embedded in CLIs this way and super appreciate how long it’s been a part of DDEV. I work for ngrok and wanted to see if we can help improve how `ddev share` invokes ngrok in a way that also doesn’t block your goal to support other tunnels.

I had two thoughts off the bat:

1. Happy to collate any feedback around how you integrated the agent and bring it back to the team. Same with our Go SDK if you’ve experimented with that at all via #4365? We’re super curious about the friction vs. benefits to users when OSS projects embed ngrok like this.
2. We’ve also introduced an entire configuration language ([Traffic Policy](https://ngrok.com/docs/traffic-policy/)) that could potentially broaden the use cases? I could help document a few options that your community might find most valuable with a touch more understanding of where ngrok feels limiting right now.

**Comment by rfay** [comment]
Wow, great to see you here @joelhans !

We haven't actually tried the go integration, I imagine it would work fine. 

Looks like "traffic policy" implements some things that the cloudflare tunnel does, could be useful. 

Our biggest problems are with CMSs that have implied URLs and won't work with ngrok URLs, see the workarounds in https://ddev.readthedocs.io/en/stable/users/topics/sharing/ - it's a silliness (IMO) of those products, I don't *think* ngrok would have a solution to it, but it would sure be interesting.

**Comment by rfay** [comment]
BTW, we're mighty glad we chose and implemented ngrok, as it's been stable and successful so many years, compared with all the other tunnels that come and go.

**Comment by atj4me** [comment]
Hello everyone, 

I just built an addon to replace ngrok for me. Maybe it could work out to some, if they use TailScale already like I do. 

If you use TailScale, it will just spin up a new container and route HTTP PORT traffic to TailScale container enabling it to be shared publicly or privately. On the tailscale end, it will create a new node with dedicated machine name ( that is based on the project name ) and accessible via MagicDNS. 

Thought I'd share here too.

**Comment by atj4me** [comment]
> [@joelhans](https://github.com/joelhans)
> 
> The issue for WordPress is that only supports absolute urls, not relative paths. I spent probably 100+ hours exploring this, implementing all sorts of regex rewrites of request and response body and headers from within a Traefik Middleware, but never fully succeeded - little edge cases kept popping up that either broke front-end functionality, or mangled the database with conflicting domains stored in various url paths. The juice was ultimately not worth the squeeze for me.
> 
> Though, I recently came across this project (https://codobee.com/) that apparently aims to offer a sort of ngrok-like service, and claims to have handled all of those rewrites. I chatted with their team and pointed out some of the foibles that I had encountered. I'm awaiting an update from them.
> 
> If ngrok were to take on the task of figuring out all of the necessary rewrites on behalf of everyone, it could be a worthwhile endeavour.
> 
> I hope this helps.

@nickchomey 
I kind of once overcome this using Apache ProxyPass after I saw it being used on a Java application with Apache Server. 

It basically proxies all requests to the webserver at https://<yourdomain> to the wordpress host! 

So the host sees a request coming to the actual url, and doesn't need URL Rewrites, while the url entered on browser is totally different one. I hope HAProxy can be implemented on top of Traefic or anywhere so one can make the Wordpress accessible in a different URL than the configured one.

**Comment by nickchomey** [comment]
Thanks, but that's not at all the problem. 

It was relatively easy to rewrite the host with traefik middlewares once I wrapped my head around how traefik works. 

The real issue is that sometimes the host is included in query strings and POST request body/payload, in inconsistent formats. I started creating rules for handling that, but it became too much work.

Also, I do believe it was necessary to also rewrite the urls contained in the html response body, rather than just rewrite the host as it came through. I don't recall why though.

But if you use WordPress and have a solution for this, please do share!

**Comment by atj4me** [comment]
@nickchomey Oh! I did this a while ago when I had deployed wordpress on an internal node and want to make it visible to external clients for demo. It worked for an extend, not without issues. 

Now, I'd rather use a single URL and when it is time to deploy, I use `wp rewrite` from cli and it would rewrite everything to new URL. Works flawlessly.

**Comment by nickchomey** [comment]
Yes, that was ultimately my conclusion. Wp has been staunchly against relative paths forever, so it's ultimately futile to swim against the tide - even if you can succeed for the most part, edge cases just keep popping up that make it not worthwhile.

Though, what I really wanted this for was to be able to use/issue different subdomains that could be shared with/accessed by others via cloudflare tunnels. I had that mostly set up as well, though in a very hacky way.

I look forward to checking out your tailscale efforts to see how it might work with all of this for WordPress!

**Comment by atj4me** [comment]
It would be kind of similar. Instead of your own domain for Cloudflare tunnel, Tailscale will have a MagicDNS on free plans. It is a set of random subdomains under .ts.net. 

And then if your project is wp1, and magic dns is yak-bebop.ts.bet, you will essentially get a wordpress site running under wp1.yak-bebop.ts.net

If you set the env to public, if will be public, otherwise only on devices authenticated by TailScale by running their VPN. 

Kind of the reason I went for this approach, because 1. I didn't have a domain and 2. I cannot change URLs on third party sites ( Stripe and Agora ) for sending webhooks, and 3. Don't want to keep running a foreground process. 

Tailscale is using WireGuard VPN under the hood, and has a granular ACL system for permissions, and it checks all of my boxes!

**Comment by rfay** [comment]
Related: https://github.com/davo20019/ddev-share-cf (cloudflare share)
